DaemonSet: Respect ControllerRef

enisoc opened the pull request:
**What this PR does / why we need it**:

This is part of the completion of the [ControllerRef](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) proposal. It brings DaemonSet into full compliance with ControllerRef. See the individual commit messages for details.

**Which issue this PR fixes**:

This ensures that DaemonSet does not fight with other controllers over control of Pods.

**Special notes for your reviewer**:

**Release note**:

```release-note
DaemonSet now respects ControllerRef to avoid fighting over Pods.
```
cc @erictune @kubernetes/sig-apps-pr-reviews

Commit:
DaemonSet: Add ControllerRef on all created Pods.

Commit:
DaemonSet: Set DefaultGarbageCollectionPolicy to OrphanDependents.

Now that DaemonSet adds ControllerRef to Pods it creates,
we need to set this default so legacy behavior is maintained.

Commit:
DaemonSet: Use ControllerRefManager to adopt/orphan.

Commit:
DaemonSet: Use ControllerRef to route watch events.

This is part of the completion of ControllerRef, as described here:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md#watches

Commit:
DaemonSet: Update Lister documentation for ControllerRef.

The DaemonSet Listers still use selectors, because this is the
behavior expected by callers. This clarifies the meaning of the
returned list. Some callers may need to switch to using
GetControllerOf() instead, but that is a separate, case-by-case issue.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42173)
<!-- Reviewable:end -->

kow3ns commented:
Reviewed 33 of 33 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/controller/daemon/daemoncontroller.go, line 73 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0F7a-G0UpKNBMEqar:-Ke0F7a-G0UpKNBMEqas:bixxmvg) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2cf9d9986d94ebb5479476a5f6078646abc610d/pkg/controller/daemon/daemoncontroller.go#L73)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = extensions.SchemeGroupVersion.WithKind("DaemonSet")
> ```

You sure you want this to be package public?

---

*[pkg/controller/daemon/daemoncontroller.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0FoZCJZcQBs_gMqNo:-Ke0FoZCJZcQBs_gMqNp:brkrjt6) ([raw file](https://github.com/kubernetes/kubernetes/blob/7f626cf8365a96d3fbdbe9510386321f7b5d0223/pkg/controller/daemon/daemoncontroller.go#L103)):*
> ```Go
> 	nodeStoreSynced cache.InformerSynced
> 
> 	lookupCache *controller.MatchingCache
> ```

Are we removing the cache because shared

---

*[pkg/controller/replicaset/replica_set.go, line 62 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0It1NseI33OEGA6KH:-Ke0It1O_9UHebvw-ZeV:b-kk6lw0) ([raw file](https://github.com/kubernetes/kubernetes/blob/f2cf9d9986d94ebb5479476a5f6078646abc610d/pkg/controller/replicaset/replica_set.go#L62)):*
> ```Go
> 
> // ControllerKind contains the schema.GroupVersionKind for this controller type.
> var ControllerKind = v1beta1.SchemeGroupVersion.WithKind("ReplicaSet")
> ```

You sure you want to export this?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
*[pkg/controller/daemon/daemoncontroller.go, line 103 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-Ke0FoZCJZcQBs_gMqNo:-Ke0JmtOWtPY-ejvAWLK:binvtqr) ([raw file](https://github.com/kubernetes/kubernetes/blob/7f626cf8365a96d3fbdbe9510386321f7b5d0223/pkg/controller/daemon/daemoncontroller.go#L103)):*
<details><summary><i>Previously, kow3ns (Kenneth Owens) wrote…</i></summary><blockquote>

Are we removing the cache because shared
</blockquote></details>

informers make it no longer necessary?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173)*
<!-- Sent from Reviewable.io -->

kow3ns commented:
Other than the comments lgtm.  Are you removing the cache entirely because the shared informers mitigate its usefulness.

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42173#-:-Ke0Oh39JlyTV6qv80IT:bcgyuz4)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented on pkg/controller/controller_utils.go:
```diff
@@ -481,13 +478,27 @@ func (r RealPodControl) CreatePodsWithControllerRef(namespace string, template *
 		return fmt.Errorf("controllerRef has empty Kind")
 	}
 	if controllerRef.Controller == nil || *controllerRef.Controller != true {
```
Could you also verify that controllerRef.BlockOwnerDeletion is true, and set it to true in all controllee's creation/adoption? See the [comments](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L259-L266) here.

enisoc replied:
I will do this in a separate PR since the change will apply to all controllers.

enisoc replied:
After looking more closely, it makes more sense to fix it here. I just needed to move the BlockOwnerDeletion check up into validateControllerRef().

caesarxuchao commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -580,7 +583,15 @@ func (dsc *DaemonSetsController) syncNodes(ds *extensions.DaemonSet, podsToDelet
 		go func(ix int) {
 			defer createWait.Done()
-			if err := dsc.podControl.CreatePodsOnNode(nodesNeedingDaemonPods[ix], ds.Namespace, &template, ds); err != nil {
+			isController := true
```
Could you set the BlockOwnerDeletion like what we did for replica set controller [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L490-L491)? Please update all for all the controllers.

enisoc replied:
Done.

Commit:
DaemonSet: Don't log Pod events unless some DaemonSet cares.

Commit:
DaemonSet: Always set BlockOwnerDeletion in ControllerRef.

kargakis commented on pkg/controller/controller_utils.go:
```diff
@@ -466,11 +467,7 @@ func getPodsPrefix(controllerName string) string {
-}
-
-func (r RealPodControl) CreatePodsWithControllerRef(namespace string, template *v1.PodTemplateSpec, controllerObject runtime.Object, controllerRef *metav1.OwnerReference) error {
+func validateControllerRef(controllerRef *metav1.OwnerReference) error {
```
Is this validation running in the api server or only in controllers?

caesarxuchao replied:
Only in controllers.

caesarxuchao replied:
This validation is only required for controllers. API server doesn't know if a request is sent by a controller thus cannot do the validation.

If a TPR creates dependents, then its custom controller could do a similar check.

kargakis commented:
Should we have it in the api server too for TPRs?

On Fri, Mar 3, 2017 at 7:26 PM, Chao Xu <notifications@github.com> wrote:

> *@caesarxuchao* commented on this pull request.
> ------------------------------
>
> In pkg/controller/controller_utils.go
> <https://github.com/kubernetes/kubernetes/pull/42173#discussion_r104213152>
> :
>
> > @@ -466,11 +467,7 @@ func getPodsPrefix(controllerName string) string {
>  	return prefix
>  }
>
> -func (r RealPodControl) CreatePods(namespace string, template *v1.PodTemplateSpec, object runtime.Object) error {
> -	return r.createPods("", namespace, template, object, nil)
> -}
> -
> -func (r RealPodControl) CreatePodsWithControllerRef(namespace string, template *v1.PodTemplateSpec, controllerObject runtime.Object, controllerRef *metav1.OwnerReference) error {
> +func validateControllerRef(controllerRef *metav1.OwnerReference) error {
>
> Only in controllers.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/42173#discussion_r104213152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf7-sMpJ9cP6WrQS1PYEmK_Ma24Ykks5riFtMgaJpZM4MNbnP>
> .
>

Commit:
DaemonSet: Check that ControllerRef UID matches.

enisoc commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this

janetkuo commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -134,7 +134,7 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet, allControllers []string, disabled
 	fs.Int32Var(&s.ConcurrentSATokenSyncs, "concurrent-serviceaccount-token-syncs", s.ConcurrentSATokenSyncs, "The number of service account token objects that are allowed to sync concurrently. Larger number = more responsive token generation, but more CPU (and network) load")
 	fs.Int32Var(&s.LookupCacheSizeForRC, "replication-controller-lookup-cache-size", s.LookupCacheSizeForRC, "This flag is deprecated and will be removed in future releases. ReplicationController no longer requires a lookup cache.")
 	fs.Int32Var(&s.LookupCacheSizeForRS, "replicaset-lookup-cache-size", s.LookupCacheSizeForRS, "This flag is deprecated and will be removed in future releases. ReplicaSet no longer requires a lookup cache.")
-	fs.Int32Var(&s.LookupCacheSizeForDaemonSet, "daemonset-lookup-cache-size", s.LookupCacheSizeForDaemonSet, "The the size of lookup cache for daemonsets. Larger number = more responsive daemonsets, but more MEM load.")
```
We should call `fs.MarkDeprecated`, and put a TODO for when to remove this flag

enisoc replied:
Thanks I fixed it here and added it to my list of fixes for RC/RS that I will make in a separate PR.

Commit:
DaemonSet: Mark daemonset-lookup-cache-size flag as deprecated.

janetkuo commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -765,14 +784,14 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 		switch ds.Spec.UpdateStrategy.Type {
 		case extensions.RollingUpdateDaemonSetStrategyType:
-			err = dsc.rollingUpdate(ds)
+			err = dsc.rollingUpdate(ds, nodeToDaemonPods)
```
We should probably add a test to catch this?

enisoc replied:
In other controllers, we expect changes we make to be observed only in the next sync, and we re-queue when appropriate. Does DaemonSet not work that way?

janetkuo replied:
No, DaemonSet controller didn't work that way. In the rollingUpdate case, we may create/kill some pods after `manage`, and then do the rolling update based on `nodeToDaemonPods`. If we use the cached value after we changed those pods in `manage`, rolling update may create or kill more pods than needed. It might be fine for Deployments, but it's safer not to do this in DaemonSet controller (since every node expects one and exactly one pod.)

enisoc replied:
OK I switched back to relisting before each phase, and added a test. PTAL.

enisoc commented on pkg/controller/controller_utils.go:
```diff
@@ -481,16 +478,30 @@ func (r RealPodControl) CreatePodsWithControllerRef(namespace string, template *
 		return fmt.Errorf("controllerRef has empty Kind")
 	}
 	if controllerRef.Controller == nil || *controllerRef.Controller != true {
```
I'll make a note to address that in a future PR. It will affect multiple controllers and may require updates to their unit tests, since not all tests bother to fill those in. In this PR, I am just moving this validation code to a common function so I can use it in two places without duplicating.

janetkuo replied:
sounds good

janetkuo commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -1167,3 +1189,262 @@ func TestUpdateNode(t *testing.T) {
+	pod := newPod("pod1-", "node-0", simpleDaemonSetLabel, nil)
+	manager.addPod(pod)
+	if got, want := manager.queue.Len(), 2; got != want {
+		t.Fatalf("queue.Len() = %v, want %v", got, want)
```
also check the items in the queue

enisoc replied:
Fixed in upcoming push to the PR.

janetkuo commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -1167,3 +1189,262 @@ func TestUpdateNode(t *testing.T) {
+	bumpResourceVersion(pod)
+	manager.updatePod(&prev, pod)
+	if got, want := manager.queue.Len(), 2; got != want {
+		t.Fatalf("queue.Len() = %v, want %v", got, want)
```
ditto

enisoc replied:
Fixed in upcoming push to the PR.

enisoc commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -273,78 +258,57 @@ func (dsc *DaemonSetsController) enqueueDaemonSetAfter(obj interface{}, after ti
+		}
 		dsKey, err := controller.KeyFunc(ds)
 		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", ds, err))
```
In another ControllerRef PR, the reviewers asked me to remove this because it's too noisy for a watch event handler. I applied the change across all my ControllerRef PRs. The sync loop is a bit different because it's specific to DaemonSet, whereas the watch even handlers get called for many events that they end up not caring about.

janetkuo replied:
sounds good

janetkuo commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -1167,3 +1189,262 @@ func TestUpdateNode(t *testing.T) {
+	}
+}
+
+func TestUpdatePodOrphanWithNewLabels(t *testing.T) {
```
`...WithMatchingLabels`

enisoc replied:
Fixed in upcoming push to the PR.

enisoc commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -462,32 +453,58 @@ func (dsc *DaemonSetsController) updateNode(old, cur interface{}) {
+	nodeToDaemonPods := make(map[string][]*v1.Pod)
+	for _, pod := range claimedPods {
+		nodeName := pod.Spec.NodeName
+		nodeToDaemonPods[nodeName] = append(nodeToDaemonPods[nodeName], pod)
```
I think that's a separate issue since I'm not changing the semantics in this PR. Note that the previous line `daemonPod := &(*daemonPods[i])` was not actually doing a copy ([example](https://play.golang.org/p/wtLp9uxU8O)).

Can we discuss this after landing ControllerRef? We should also think about the general pattern to use across controllers.

janetkuo replied:
Yes we can address this later

kow3ns replied:
StatefulSet is now deep copying everything FWIW.

janetkuo commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -1167,3 +1189,262 @@ func TestUpdateNode(t *testing.T) {
+	prev := *pod
+	pod.OwnerReferences = nil
+	bumpResourceVersion(pod)
+	manager.updatePod(&prev, pod)
```
Isn't this adopt rather than release?

enisoc replied:
The name might be confusing. It is testing the case that a pod *was* released and we are observing the update event. You're right that we are expecting the reaction to be "queue for potential adoption".

janetkuo replied:
The new name looks good

enisoc commented on pkg/controller/daemon/daemoncontroller_test.go:
```diff
@@ -269,6 +274,22 @@ func validateSyncDaemonSets(t *testing.T, fakePodControl *fakePodControl, expect
+		if got, want := controllerRef.APIVersion, "extensions/v1beta1"; got != want {
+			t.Errorf("controllerRef.APIVersion = %q, want %q", got, want)
+		}
+		if got, want := controllerRef.Kind, "DaemonSet"; got != want {
```
In general, it's good practice to make unit test expected values hard-coded whenever possible. This helps to avoid problems being hidden due to tests that are actually tautologies: "check that A == B but actually both come from C, so the test passes even if C is wrong."

In particular, having hard-coded values in this test saved me in the CronJob PR when I rebased and someone had renamed a package on import. That actually made `controllerKind` use the wrong `SchemeGroupVersion`, but this test caught it.

janetkuo replied:
Thanks for the explanation

enisoc commented:
@k8s-bot gci gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot gce etcd3 e2e test this

Commit:
DaemonSet: Relist Pods before each phase of sync.

The design of DaemonSet requires a relist before each phase (manage,
update, status) because it does not short-circuit and requeue for each
action triggered.

caesarxuchao commented:
/approve

Approve the changes to client.

janetkuo commented:
LGTM. Maybe you want to squash commits first?

enisoc commented:
I usually retain separate commits that represent discrete "layers" of a PR as described here: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md#8-fix-feedback-in-a-new-commit

janetkuo commented on cmd/kube-controller-manager/app/options/options.go:
```diff
@@ -134,7 +133,8 @@ func (s *CMServer) AddFlags(fs *pflag.FlagSet, allControllers []string, disabled
 	fs.Int32Var(&s.ConcurrentSATokenSyncs, "concurrent-serviceaccount-token-syncs", s.ConcurrentSATokenSyncs, "The number of service account token objects that are allowed to sync concurrently. Larger number = more responsive token generation, but more CPU (and network) load")
 	fs.Int32Var(&s.LookupCacheSizeForRC, "replication-controller-lookup-cache-size", s.LookupCacheSizeForRC, "This flag is deprecated and will be removed in future releases. ReplicationController no longer requires a lookup cache.")
 	fs.Int32Var(&s.LookupCacheSizeForRS, "replicaset-lookup-cache-size", s.LookupCacheSizeForRS, "This flag is deprecated and will be removed in future releases. ReplicaSet no longer requires a lookup cache.")
-	fs.Int32Var(&s.LookupCacheSizeForDaemonSet, "daemonset-lookup-cache-size", s.LookupCacheSizeForDaemonSet, "The the size of lookup cache for daemonsets. Larger number = more responsive daemonsets, but more MEM load.")
```
Copying from my https://github.com/kubernetes/kubernetes/pull/42659/files#r104817545:

> `MarkDeprecated` prints deprecation warnings for you, but we still need these flags.

enisoc replied:
Done.

janetkuo commented:
One last comment and LGTM otherwise

janetkuo commented:
/approve
/lgtm

lavalamp commented:
/approve

(for pkg/registry, componentconfig, and kube-controller-manager changes only)

enisoc commented:
@k8s-bot bazel test this
@k8s-bot gci gce e2e test this

liggitt commented:
was there agreement that the controller ref proposal would be implemented post-code-freeze? this proposal seems to involve rewriting all the controllers. these are significant changes dropping extremely late in the release.

enisoc commented:
@liggitt You're right that I was late, and I make no excuses for that. I was asked to keep working on this because the goal is an important stability improvement with the potential to significantly lessen the pager load for us over the next quarter and reduce confusing breakages for users.

Given that, we classified this as bugfix/stability work, and I tried to follow the procedures for that. Although the devil is always in the details, we think these changes are low risk relative to the known problems they address, because the insertion points are clearly defined and largely orthogonal to the primary business logic of controllers.

In particular, the main insertion points are:

1. Apply ControllerRef to claim ownership of objects you create.
2. When listing objects to manage, filter out anything you don't own.
3. When a watch event is received, route it to the owning controller based on ControllerRef.

Points 1 and 2 were already implemented for RC and RS in v1.5, and I have now copied that pattern to the other controllers. Point 3 fixes a remaining bug that could result in RC/RS stalling and never syncing. The real meat of the controllers is in the actual managing of those objects once they're found, and I haven't touched those parts.

If you think this needs to be discussed with a larger audience, please include me, @janetkuo and @erictune in your communication. Thanks!

lavalamp commented:
I added do-not-merge to the others in this series until we resolve this.

lavalamp commented:
FWIW, I think we really need these changes, they are blocking other improvements.

liggitt commented:
I agree fighting controllers is bad, but these PRs are the definition of "planned bug fixes", which we've explicitly said we weren't doing.

Process aside, I just want a stable release. These rewrites are big enough that I can't tell whether they will help with a stable release or not. We know we have problems with fighting controllers, but these could be trading those for unknown problems with orphaning, etc.

In general, I'm in favor of the changes, but I'm concerned we're losing all our soak time for 1.6. At the very least, the decision to make these changes to complete a proposal should have been done transparently via the exception process.

erictune commented:
@ethernetdan   there is a difference of opinion on whether this PR was a "just a bug fix" and thus able to go in after the freeze, or whether it changes enough that it is a feature, and thus should not have merged after the freeze (it was only lgmt'ed after the freeze).

lavalamp commented:
(and #42175, #42176, #42177, #42659 are all in the same boat)

liggitt commented:
The "this is part of completing a proposal" comment really makes it seem like a feature, albeit one intended to stabilize and improve the controllers. 

Either way, we should decide quickly. If it is the judgement of the relevant owners and release leads that the benefit outweighs the risk, that's great, and we should get what soak time we can. Maybe that discussion already happened, there was just no visibility to it.

ethernetdan commented:
> @ethernetdan there is a difference of opinion on whether this PR was a "just a bug fix" and thus able to go in after the freeze, or whether it changes enough that it is a feature, and thus should not have merged after the freeze (it was only lgmt'ed after the freeze).

One of the takeaways from this process is that we need to have clearer guidelines on constitutes a bug for code freeze. With the size of these changes it probably should have gone through the exception process and were other similar changes that did not have their exceptions accepted.

> Either way, we should decide quickly. If it is the judgement of the relevant owners and release leads that the benefit outweighs the risk, that's great, and we should get what soak time we can. Maybe that discussion already happened, there was just no visibility to it.

That discussion hasn't occurred yet. My current thinking given where we are in the release process is we move in the direction that least jeopardizes the release:
* If we feel confident in the rest of the PRs, we continue moving forward with them
* If we don't feel confident in either the remaining PRs or the ones that merged, we should revert

Regardless we should try to move quickly here.
Could all relevant reviewers/owners please weigh in:  @enisoc,@kargakis, @kow3ns, @janetkuo, @lukaszo, @caesarxuchao. Let's try to make a decision by EOD if possible.

enisoc commented:
@ethernetdan Just to clarify, the remaining unmerged PRs are independent of the ones that merged, so a third option is to stop here and neither merge nor revert anything.

As for weighing in, the summary of my opinion in https://github.com/kubernetes/kubernetes/pull/42173#issuecomment-284965630 is that the risk is low because RC/RS were already doing this since v1.5. If we stop with what has merged so far, the net result is that in v1.6 we will roll out that pattern to a few more controllers: SS, DS, Deployment. As I understand it, these are the ones that are causing the most user/ops pain currently.

ethernetdan commented:
@enisoc thanks for the clarification. In that case I would propose that we keep what's merged and hold off on others until 1.7 - unless they're any objections?

kargakis commented:
Fine by me

liggitt commented:
It appears the ones already merged are causing issues with garbage collection: https://github.com/kubernetes/kubernetes/issues/42639

enisoc commented:
We are discussing the GC issue on https://github.com/kubernetes/kubernetes/issues/42639, but for anyone following here: That flake appears to be due to a known race condition (item 6 in https://github.com/kubernetes/kubernetes/issues/26120) that predates my PRs and already exists in v1.5 for RC/RS.
